bash_completionsV2.go

```diff
@@ -230,43 +218,47 @@ __%[1]s_handle_special_char()
 __%[1]s_format_comp_descriptions()
 {
     local tab=$'\t'
-    local comp="$1"
-    local longest=$2
-
-    # Properly format the description string which follows a tab character if there is one
-    if [[ "$comp" == *$tab* ]]; then
-        desc=${comp#*$tab}
-        comp=${comp%%%%$tab*}
-
-        # $COLUMNS stores the current shell width.
-        # Remove an extra 4 because we add 2 spaces and 2 parentheses.
-        maxdesclength=$(( COLUMNS - longest - 4 ))
-
-        # Make sure we can fit a description of at least 8 characters
-        # if we are to align the descriptions.
-        if [[ $maxdesclength -gt 8 ]]; then
-            # Add the proper number of spaces to align the descriptions
-            for ((i = ${#comp} ; i < longest ; i++)); do
-                comp+=" "
-            done
-        else
-            # Don't pad the descriptions so we can fit more text after the completion
-            maxdesclength=$(( COLUMNS - ${#comp} - 4 ))
-        fi
+    local comp desc maxdesclength
+    local longest=$1
+
+    local i ci
+    for ci in ${!COMPREPLY[*]}; do
```

Comment:
I wasn't familiar with this syntax.  From what I understand it loops over the array but put the index in $ci.
For my education why is that better than something like `for ((ci = 0 ; ci < ${#COMPREPLY} ; ci++)); do` ?

When trying it I notice that $COMPREPLY has a bunch of empty elements which your syntax ignores automatically.  Is that why you choose that syntax?

Reply:
The niceties of that syntax boil down to that it excludes unassigned indices. Sparse arrays "just work" (if they work in the context at hand in other respects in the first place), no need to worry about accessing unset data in `nounset` mode or to wonder if all desired items are actually iterated over, for example:

```shellsession
$ shopt -po nounset
set -o nounset
$ a=(1 2 3)
$ unset 'a[1]'
$ for i in ${!a[*]}; do echo "${a[i]}"; done
1
3
$ for ((i = 0; i < ${#a[*]}; i++)); do echo "${a[i]}"; done
1
bash: a[i]: unbound variable
# oops!

$ shopt -po nounset
set +o nounset
$ for ((i = 0; i < ${#a[*]}; i++)); do echo "${a[i]}"; done
1

# oops!
```

I suppose in this particular case it's not that likely that `COMPREPLY` would actually contain unassigned indices, so it doesn't matter that much which approach would be used. But I find the syntax I used here more succinct, and if one can rely on it being used to iterate over arrays, the convenience of being able to remove array elements by simply `unset`ing them like in the example above is a nice bonus.

Also, that's the syntax for iterating over associative array keys, so there's a consistency bonus attached as well.

Reply:
Thanks for the explanation. I leaned something new 👍